lib/srv/db/elasticsearch/engine.go

```diff
@@ -0,0 +1,207 @@
+	}{
+		Success: false,
+		Message: err.Error(),
+	})
+	if err != nil {
+		e.Log.WithError(err).Error("failed to marshal error response")
+		return
+	}
+
+	response := &http.Response{
+		ProtoMajor: 1,
+		ProtoMinor: 1,
+		StatusCode: statusCode,
+		Body:       io.NopCloser(bytes.NewBuffer(jsonBody)),
+		Header: map[string][]string{
+			"Content-Type":   {"application/json"},
+			"Content-Length": {strconv.Itoa(len(jsonBody))},
+		},
+	}
+
+	if err := response.Write(e.clientConn); err != nil {
+		e.Log.Errorf("elasticsearch error: %+v", trace.Unwrap(err))
+		return
+	}
+}
+
+func (e *Engine) HandleConnection(ctx context.Context, sessionCtx *common.Session) error {
+	if err := e.authorizeConnection(ctx); err != nil {
+		return trace.Wrap(err)
+	}
+
+	e.Audit.OnSessionStart(e.Context, sessionCtx, nil)
+	defer e.Audit.OnSessionEnd(e.Context, sessionCtx)
+
+	clientConnReader := bufio.NewReader(e.clientConn)
+
+	for {
+		req, err := http.ReadRequest(clientConnReader)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		err = e.process(ctx, sessionCtx, req)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+}
+
+func copyRequest(ctx context.Context, req *http.Request) (*http.Request, error) {
+	reqCopy, err := http.NewRequestWithContext(ctx, req.Method, req.URL.String(), req.Body)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	reqCopy.Header = req.Header.Clone()
+
+	return reqCopy, nil
+}
+
+// process reads request from connected elasticsearch client, processes the requests/responses and send data back
+// to the client.
+func (e *Engine) process(ctx context.Context, sessionCtx *common.Session, req *http.Request) error {
+	reqCopy, err := copyRequest(ctx, req)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// force HTTPS, set host URL.
+	reqCopy.URL.Scheme = "https"
+	reqCopy.URL.Host = sessionCtx.Database.GetURI()
+
+	tlsConfig, err := e.Auth.GetTLSConfig(ctx, sessionCtx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: tlsConfig,
+		},
+	}
+
+	// Send the request to elasticsearch API
+	resp, err := client.Do(reqCopy)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer resp.Body.Close()
+
+	return trace.Wrap(e.sendResponse(resp))
+}
+
+// sendResponse sends the response back to the elasticsearch client.
+func (e *Engine) sendResponse(resp *http.Response) error {
+	if err := resp.Write(e.clientConn); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// authorizeConnection does authorization check for elasticsearch connection about
+// to be established.
+func (e *Engine) authorizeConnection(ctx context.Context) error {
+	ap, err := e.Auth.GetAuthPreference(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	mfaParams := services.AccessMFAParams{
+		Verified:       e.sessionCtx.Identity.MFAVerified != "",
+		AlwaysRequired: ap.GetRequireSessionMFA(),
+	}
+
+	dbRoleMatchers := role.DatabaseRoleMatchers(
```

Comment:
DatabaseRoleMatchers will return  the default access scheme that will validate user and databaseName.

Reply:
True. Is this a problem?

Reply:
I realized we don't really support db/index restrictions for Elasticsearch, so we should probably ignore actual db name - same as for other dbs with similar limitations.